src/include/server/mir/shell/surface_ready_registrar.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright © 2015 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_SHELL_SURFACE_READY_REGISTRAR_H_
+#define MIR_SHELL_SURFACE_READY_REGISTRAR_H_
+
+#include <functional>
+#include <memory>
+#include <map>
+#include <mutex>
+
+namespace mir
+{
+namespace scene { class Session; class Surface; }
+namespace geometry { struct Size; }
+
+namespace shell
+{
+/// Notifies the user when a surface is ready
+class SurfaceReadyRegistrar
+{
+public:
+    using ActivateFunction = std::function<void(std::shared_ptr<scene::Surface> const& surface)>;
+
+    SurfaceReadyRegistrar(ActivateFunction const& surface_ready);
+    ~SurfaceReadyRegistrar();
+    void register_surface(std::shared_ptr<scene::Surface> const& surface);
```

Comment:
What are the requirements on callers to ensure that this doesn't race? That is: what do callers need to do to guarantee that `surface_ready` will not miss the first buffer being submitted?